### ISSUE TITLE
TfLite Quantization Failure: Array dense_prev, which is an input to the Concatenation operator producing the output array dense_input, is lacking min/max data,

### ISSUE THREAD

**Issue by abladdha** [tensorflow/tensorflow#22528]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.11.0-rc1
- **Python version**: 2.7

### Describe the problem
I am trying to save Quantized version of graph in tflite format but getting the following error

F tensorflow/contrib/lite/toco/tooling_util.cc:1633] Array dense_prev, which is an input to the Concatenation operator producing the output array dense_input, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.

I am not able to understand why isn't Min/Max information stored for dense_prev operator if for dense_typed it's not giving any error.  

The code to reproduce the problem  
```
import tensorflow as tf
from tensorflow.contrib.lite.toco import types_pb2 as _types_pb2
QUANTIZED_UINT8 = _types_pb2.QUANTIZED_UINT8
import numpy as np

def train_model(vocab_typed_size,vocab_prev_size, decoder_size, vocab_output_size):
    g = tf.Graph()
    with tf.Session(graph=g) as sess:

        data_x = tf.placeholder(tf.float32, shape=(vocab_typed_size + vocab_prev_size), name="data_x" )
        label = tf.placeholder(tf.float32, shape=(vocab_output_size), name='labels')

        data_x_typed = data_x[:vocab_typed_size]
        data_x_prev = data_x[vocab_typed_size:]
        
        embedding_typed = tf.get_variable(name="embedding_typed",shape=(vocab_typed_size, 100))
        embedding_prev = tf.get_variable(name="embedding_prev",shape=(vocab_prev_size, 100))

        dense_typed = tf.matmul(tf.reshape(data_x_typed, (1,vocab_typed_size)),embedding_typed, name="dense_typed")
        dense_prev = tf.matmul(tf.reshape(data_x_prev, (1,vocab_prev_size)),embedding_prev,name="dense_prev")
        dense_input = tf.concat([dense_prev,dense_typed], axis = 1,name="dense_input")

        h2_w = tf.get_variable(shape=(200 , decoder_size),name="h2_w")
        h2_b = tf.get_variable(shape=(1,decoder_size),name="h2_b")
        h2_out = tf.add(tf.matmul(dense_input, h2_w), h2_b, name ="h2_out")

        decoder_w = tf.get_variable(shape=[decoder_size,vocab_output_size],name="decoder_w")     
        decoder_b = tf.get_variable(shape=[1,vocab_output_size],name="decoder_b")

        logits_message = tf.add(tf.matmul(tf.nn.relu(h2_out), decoder_w),decoder_b, name='logits_message')

        probs_message = tf.nn.sigmoid(logits_message, name = 'probs_sticker')

        # Loss function and optimizer
        loss = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits=logits_message, labels=label), name="loss")

        # Call the training rewrite which rewrites the graph in-place with
        # FakeQuantization nodes and folds batchnorm for training. It is
        # often needed to fine tune a floating point model for quantization
        # with this training tool. When training from scratch, quant_delay
        # can be used to activate quantization after training to converge
        # with the float graph, effectively fine-tuning the model.

        tf.contrib.quantize.create_training_graph(quant_delay=0)

        optimizer = tf.train.AdamOptimizer().minimize(loss)

        sess.run(tf.global_variables_initializer())
        saver = tf.train.Saver()
        print "Dumping model"
        saver.save(sess, './test/oink.ckpt')

        writer = tf.summary.FileWriter("./test/", sess.graph)
        writer.close()
        

def eval_model(vocab_typed_size,vocab_prev_size, decoder_size, vocab_output_size):
    g = tf.Graph()
    with tf.Session(graph=g) as sess:
        data_x = tf.placeholder(tf.float32, shape=(vocab_typed_size + vocab_prev_size), name="data_x" )
        data_x_typed = data_x[:vocab_typed_size]
        data_x_prev = data_x[vocab_typed_size:]

        embedding_typed = tf.get_variable(name="embedding_typed",shape=(vocab_typed_size, 100))
        embedding_prev = tf.get_variable(name="embedding_prev",shape=(vocab_prev_size, 100))

        dense_typed = tf.matmul(tf.reshape(data_x_typed, (1,vocab_typed_size)),embedding_typed, name="dense_typed")
        dense_prev = tf.matmul(tf.reshape(data_x_prev, (1,vocab_prev_size)),embedding_prev,name="dense_prev")
        
        dense_input = tf.concat([dense_prev,dense_typed], axis = 1,name="dense_input")

        h2_w = tf.get_variable(shape=(200 , decoder_size),name="h2_w")
        h2_b = tf.get_variable(shape=(1,decoder_size),name="h2_b")
        h2_out = tf.add(tf.matmul(dense_input, h2_w), h2_b, name ="h2_out")

        decoder_w = tf.get_variable(shape=[decoder_size,vocab_output_size],name="decoder_w")     
        decoder_b = tf.get_variable(shape=[1,vocab_output_size],name="decoder_b")

        logits_message = tf.add(tf.matmul(tf.nn.relu(h2_out), decoder_w),decoder_b, name='logits_message')
        probs_message = tf.nn.sigmoid(logits_message, name = 'probs_sticker')

        tf.contrib.quantize.create_eval_graph()

        saver = tf.train.Saver()
        saver.restore(sess, "./test/oink.ckpt")

        converter = tf.contrib.lite.TocoConverter.from_session(sess, [data_x], [probs_message])
        converter.quantize_weights = True
        converter.inference_type = tf.contrib.lite.constants.QUANTIZED_UINT8
        input_arrays = converter.get_input_arrays()
        print (input_arrays)
        converter.quantized_input_stats = {input_arrays[0] : (0, 2)}
        #converter.default_ranges_stats = (-1.0,1.0)
        converter.change_concat_input_ranges = True
        tflite_model = converter.convert()
        open("sr.tflite", "wb").write(tflite_model)

if __name__ == "__main__":  
    train_model(50000, 50000, 100, 10000)
    eval_model(50000, 50000, 100, 10000)
 ```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by wt-huang** [comment]
@abladdha It appears that your model is word2vec based, whose quantization task is not supported at the moment but will be in the future. Will keep you posted on the progress.

**Comment by abladdha** [comment]
@wt-huang I agree that word2vec based approaches are not yet supported. But the operations I am using for model are all supported for the quantization in the current version of tensorflow (not using embedding lookup or LSTMs) ? If i am using all the quantized supported operations in my graph then it should be able to do qunatization or could you elaborate more on what type of combination of operators are allowed in network...

**Comment by wt-huang** [comment]
@abladdha For one, `quantized_weights` is not supported but it is in working progress.

**Comment by johncf** [comment]
@wt-huang I don't understand what you're referring to by `quantized_weights` with regards to the code given. But I have the same problem, and was wondering whether matmul operation is supported at all for quantization.

If it is indeed supported at the moment, could you explain why that particular error is being thrown in this particular case? The error being: `Array dense_prev, which is an input to the Concatenation operator producing the output array dense_input, is lacking min/max data, which is necessary for quantization.` ... where `dense_prev` is a matmul.

**Comment by wt-huang** [comment]
@johncf `quantized_weights` is not support in tflite so it would impact `matmul`. Currently this issue has been acknowledged and improvements are in progress.

**Comment by suharshs** [comment]
Hi @johncf , please use converter.post_training_quantize=True (formerly quantize_weights)

Also, set your inference type to Float for post_training_quantize=True. This is because the edges in the graph are still float for this mode, but the weights are quantized and the kernels dynamically quantize inputs.

**Comment by wt-huang** [comment]
Closing this issue, feel free to reopen if any error comes up.

**Comment by johncf** [comment]
Closing this issue because proper quantization with training is unsupported? Or did someone already fix the issue raised here in nightly?

Post-training quantization is not what this issue is about.

**Comment by suharshs** [comment]
Hi John, right now we do not have a good solution for any arbritrary model for quantized training. For graphs with Concatenation (in your case) you will need to add FakeQuant nodes (with min max variables) before and after the Concat during training (like what the contrib/quantize rewrtier tries to do) for things to work. That is why we recommend doing post_training_quantize and only resort to trying quantized training if that doesn't work well (also let us know if that doesn't work well )

We are also working on better quantization tools for fully integer quantization and will keep you posted.

**Comment by aluds123** [comment]
@suharshs  Can you guide me how to use "tf.contrib.quantize.create_training_graph()" to fine tune from a floating point checkpoint?
I alredy trained a floating point checkpoint, but when I use it to be a pretrained weight and train with "tf.contrib.quantize.create_training_graph()", I meet some problems, like the following:

![00000000](https://user-images.githubusercontent.com/43241182/78626119-5f87a300-78c0-11ea-951d-d8b7cbff8f91.PNG)
How can I solve this problem? Thank you.

**Comment by suharshs** [comment]
Hi @aluds123 , 

The contrib.quantize framework creates new TF variables to store min/max state for various tensors. When restoring from a float checkpoint, these values will not be initialized. You need to modify your restore your checkpoint and then initialize_all_variables to fill in default inital values for these remaining tf.variables.

**Comment by aluds123** [comment]
Hi @suharshs ,

I find a function like this "slim.assign_from_checkpoint_fn()", it has a parameter "ignore_missing_vars=True". I use it to import floating point checkpoint and train with "tf.contrib.quantize.create_training_graph()", and it works.
Is this use possible?
Sorry my poor English.
